btree/btree.c

```diff
@@ -2952,6 +2961,518 @@ static void ff_node_free_credits(const struct nd *node,
+	 *        value based on the index provided.
+	 *        vkv_key() and vkv_val() functions should be used to get the
+	 *        memory address pointing to key and val.
+	 *        vkv_keysize() and vkv_valsize() functions should be used to
+	 *        fill the v_count parameter of the slot vector.
+	 */
+}
+
+static void vkv_node_key(struct slot *slot)
+{
+	/**
+	 * @brief This function will fill the provided slot with its key and
+	 *        value based on the index provided.
+	 *        vkv_key() functions should be used to get the
+	 *        memory address pointing to key and val.
+	 *        vkv_keysize() functions should be used to
+	 *        fill the v_count parameter of the slot vector.
+	 */
+}
+
+static void vkv_child(struct slot *slot, struct segaddr *addr)
+{
+	/**
+	 * @brief This function will return the memory address pointing to its
+	 *        child node. We will call the function vkv_val() for this
+	 *        purpose.
+	 *        This function is called for internal nodes.
+	 */
+}
+
+static bool vkv_isfit(struct slot *slot)
+{
+	/**
+	 * @brief This function will determine if the current key and value
+	 *        can be added to the node.
+	 *        It also checks the validity of the record.
+	 *        Modifications will be required in nt_isfit() function
+	 *        declaration as we need to have the sizes of incoming key and
+	 *        value.
+	 *
+	 *        keysize + valsize <= vkv_space(slot->s_node)
+	 */
+}
+
+static void vkv_done(struct slot *slot, struct m0_be_tx *tx, bool modified)
+{
+
+}
+
+static void vkv_make(struct slot *slot, struct m0_be_tx *tx)
+{
+	/**
+	 * @brief This function will create space to add new entry at the
+	 *        given index. It is assumed that the possibility whether this
+	 *        can occur or not is determined before calling this function.
+	 *
+	 *        start_key_addr = vkv_key(slot->s_node, index)
+	 *        start_val_addr = vkv_val(slot->s_node, index)
+	 *        count = num of records i.e. vkv_used
+	 *
+	 *        idx = index
+	 *        while(idx <= count) {
+	 *          total_key_size += vkv_keysize(slot->s_node, idx)
+	 *          total_val_size += vkv_valsize(slot->s_node, idx)
+	 *          idx++
+	 *        }
+	 *
+	 *        //Update directory
+	 *        idx = index
+	 *        temp_count = count
+	 *        while(temp_count >= idx) {
+	 *          dir_rec[temp_count+1].key_offset = dir_rec[temp_count].key_offset
+	 *          dir_rec[temp_count+1].val_offset = dir_rec[temp_count].val_offset
+	 * 
+	 *          dir_rec[temp_count+1].key_offset = dir_rec[temp_count+1].key_offset + incoming_key_size
+	 *          dir_rec[temp_count+1].val_offset = dir_rec[temp_count+1].val_offset + incoming_value_size
+	 *          temp_count --
+	 *        }
+	 *        
+	 *        memmove(start_key_addr + incoming_key_size,start_key_addr,total_key_size)
+	 *        memmove(start_val_addr + incoming_val_size,start_val_addr,total_val_size)
```

Comment:
memmove(start_val_addr - incoming_val_size,start_val_addr,total_val_size)

here start_val_address will be address of value present at idx-1. Please check once.

Reply:
Updated.